### ISSUE TITLE
[feature request] [discussion] generalized gradient clipping

### ISSUE THREAD

**Issue by jamestwebber** [pyro-ppl/pyro#2109]
As @martinjankowiak said in the discussion of #1293, it'd be nice to have a generic way to clip gradients in Pyro. I _think_ I understand how to implement this and I'm interested in contributing the PR if I can get a little guidance. The architecture of Pyro is complex and it's not obvious to me where this feature should live.

The current way to clip gradients in PyTorch is with [`clip_grad_norm_`](https://pytorch.org/docs/stable/nn.html#clip-grad-norm) and [`clip_grad_value_`](https://pytorch.org/docs/stable/nn.html#clip-grad-value). This is different from how Pyro has implemented ClippedAdam, as the clipping happens outside of the optimizer step (see [this SO discussion](https://stackoverflow.com/questions/54716377/how-to-do-gradient-clipping-in-pytorch)):

```python
optimizer.zero_grad()
loss, hidden = model(data, hidden, targets)
loss.backward()

torch.nn.utils.clip_grad_norm_(model.parameters(), args.clip)
optimizer.step()
```

Because Pyro encapsulates those steps into e.g. the `SVI.step` method, it's not totally clear where to add this ability, but maybe the most natural place is in the `PyroOptim` class, allowing any wrapped optimizer to clip gradients if desired. It could be a global attribute of the optimizer:

```
def __call__(self, params,  *args, **kwargs):
        ...
        for p in params:
            ...

            if isinstance(self.optim_objs[p], torch.optim.lr_scheduler._LRScheduler) or \
                    isinstance(self.optim_objs[p], torch.optim.lr_scheduler.ReduceLROnPlateau):
                # if optim object was a scheduler, perform an optimizer step
                if self.clip_norm > 0:
                    clip_grad_norm_(self.optim_objs[p].optimizer.params(), self.clip_norm)
                self.optim_objs[p].optimizer.step(*args, **kwargs)
            else:
                if self.clip_norm > 0:
                    clip_grad_norm_(self.optim_objs[p].params(), self.clip_norm)
                self.optim_objs[p].step(*args, **kwargs)
```

This would entail adding `clip_norm` (or more generically `**kwargs`) to the PyroOptim init method and exposing it in e.g. `pytorch_optimizers.py` (I haven't pored over enough of the codebase to know everything that would be affected).

Doing it on a per-parameter basis might be trickier, depending on what the desired API looks like. I think I'd try to mimic the behavior of `optim_args` and write it such that `clip_norm` (or `clip_value`) is either a float or a dictionary that returns floats, and add a `_get_clip_args` method in the same way that `_get_optim_args` gets per-parameter args for the optimizer.

But obviously, I don't know the long-term plans for Pyro and this might not be the best way to implement this feature (if it's even desired). If that plan sounds reasonable I will write a PR and put it up for review, but before that I wanted to start the discussion with you all to get feedback.

**Comment by fritzo** [comment]
Sounds reasonable. `pyro.optim` indeed seems like the appropriate place. @martinjankowiak or @jpchen may have more precise opinions.

**Comment by jamestwebber** [comment]
Closed by #2111 ?
